src/wrappers/themis/wasm/src/secure_cell_context_imprint.js

```diff
@@ -0,0 +1,190 @@
+// limitations under the License.
+
+/**
+ * @file
+ * Themis Secure Cell (Context Imprint mode).
+ */
+
+const libthemis = require('./libthemis.js')
+const errors = require('./themis_error.js')
+const utils = require('./utils.js')
+
+const cryptosystem_name = 'SecureCellContextImprint'
+
+const ThemisError = errors.ThemisError
+const ThemisErrorCode = errors.ThemisErrorCode
+
+module.exports = class SecureCellContextImprint {
+    constructor(masterKey) {
+        masterKey = utils.coerceToBytes(masterKey)
+        if (masterKey.length == 0) {
+            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
+                'master key must be not empty')
+        }
+        this.masterKey = masterKey
+    }
+
+    encrypt(message, context) {
+        message = utils.coerceToBytes(message)
+        if (message.length == 0) {
+            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
+                'message must be not empty')
+        }
+
+        // Other Secure Cell kinds have optional context, so the users are likely to omit
+        // the context here as well. Let's produce a more helpful error message instead
+        // of an error about 'undefined' being invalid byte buffer type.
+        if (context === undefined) {
+            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
+                'SecureCellContextImprint requires context for encrypting')
+        }
+
+        context = utils.coerceToBytes(context)
+        if (context.length == 0) {
+            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
+                'context must be not empty')
+        }
+
+        let status
+        /// C API uses "size_t" for lengths, it's defined as "i32" in Emscripten
+        let result_length_ptr = libthemis.allocate(4, 'i32', libthemis.ALLOC_STACK)
+        let master_key_ptr, message_ptr, context_ptr, result_ptr, result_length
+        try {
+            master_key_ptr = libthemis._malloc(this.masterKey.length)
+            message_ptr = libthemis._malloc(message.length)
+            context_ptr = libthemis._malloc(context.length)
+            if (!master_key_ptr || !message_ptr || !context_ptr) {
+                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
+            }
+
+            libthemis.writeArrayToMemory(this.masterKey, master_key_ptr)
+            libthemis.writeArrayToMemory(message, message_ptr)
+            libthemis.writeArrayToMemory(context, context_ptr)
+
+            status = libthemis._themis_secure_cell_encrypt_context_imprint(
+                master_key_ptr, this.masterKey.length,
+                message_ptr, message.length,
+                context_ptr, context.length,
+                null, result_length_ptr
+            )
+            if (status != ThemisErrorCode.BUFFER_TOO_SMALL) {
+                throw new ThemisError(cryptosystem_name, status)
```

Comment:
shouldn't we clean up master key memory in case of error?

Reply:
Master key will be erased from Emscripten memory in the **finally** block below:

https://github.com/cossacklabs/themis/blob/3b00358e62863a4915795673ab3b61f0e976c122/src/wrappers/themis/wasm/src/secure_cell_context_imprint.js#L106-L112

Actually... I think that we'd also need to clean out `message` and maybe `context` as well. And the same thing for decryption: wipe the decrypted result out of Emscripted memory after we have copied it to JavaScript.